tests/testthat/test-return_linter.R

```diff
@@ -0,0 +1,842 @@
+      message = rex::rex("Use implicit return behavior; explicit return() is not needed.")
+    ),
+    return_linter()
+  )
+})
+
+test_that("Do not lint if/else statements (with return) on end of function", {
+  linter <- return_linter(use_implicit_returns = FALSE)
+
+  expect_lint(
+    trim_some("
+      function() {
+        if (x) {
+          return(4)
+        } else if (y) {
+          return(5)
+        } else {
+          return(6)
+        }
+      }
+    "),
+    NULL,
+    linter
+  )
+})
+
+test_that("Lint control statements (without return) on end of function", {
+  linter <- return_linter(use_implicit_returns = FALSE)
+  lint_msg <- rex::rex("All functions must have an explicit return().")
+
+  expect_lint(
+    trim_some("
+      function() {
+        while (x > 4) {
+          cat(4)
+          if (x < 4) {
+            return(x)
+          }
+        }
+      }
+    "),
+    list(lint_msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        repeat {
+          cat(4)
+        }
+      }
+    "),
+    list(lint_msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        for (i in 1:10) {
+          cat(4)
+          if (i > 11) {
+            return(x)
+          }
+        }
+      }
+    "),
+    list(lint_msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        if (x == 2L){
+          return(e)
+        } else if (x == 3L) {
+          cat(f)
+        }
+      }
+    "),
+    list(lint_msg, line_number = 4L),
+    linter
+  )
+})
+
+test_that("Do not lint stop on end of function", {
+  expect_lint(
+    trim_some("
+      function() {
+        # Test
+        3 + 4
+        stop(1)
+      }
+    "),
+    NULL,
+    return_linter(use_implicit_returns = FALSE)
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        stop(1)
+      }
+    "),
+    NULL,
+    return_linter()
+  )
+})
+
+test_that("Do not lint stop on end of function", {
+  linter <- return_linter(use_implicit_returns = FALSE)
+  lint_msg <- rex::rex("All functions must have an explicit return().")
+
+  expect_lint(
+    trim_some("
+      function(x) {
+        switch(x, a = 1, 'b' = 2, '3' = 3, 4)
+      }
+    "),
+    list(lint_msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function(x) {
+        switch(x, a = return(1), 'b' = stop(2), '3' = return(3), 4)
+      }
+    "),
+    list(lint_msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        switch(
+          x,
+          a = return(1),
+          'b' = stop(2),
+          '3' = return(3)
+        )
+      }
+    "),
+    list(lint_msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function(x) {
+        switch(x, a = return(1), 'b' = stop(2), '3' = return(3), stop('End'))
+      }
+    "),
+    list(lint_msg, line_number = 2L),
```

Comment:
@MichaelChirico I don't think we should give a lint here. In my opinion this is a false positive, as there is no way an implicit return is given.
My first implementation handled this accordingly. Maybe we could do that again.

Reply:
I just think using `return()` inside `switch()` is bad practice in general. In this simple case it's easy enough to read, but more practically `switch()` statements are more complicated and it will be easy to lose track of what actually gets returned. Moving `return()` to the outside of the `switch()` seems the safest approach IMO. The downside is that it's somewhat confusing to parse `return(stop('...'))`, but I think practically speaking a `stop()` is only likely to show up in the catch-all final position.

Reply:
I agree that this is bad style, but I would still like to prevent as many false positives as possible and I think that's part of it, isn't it?

Reply:
that makes sense, though I don't think it's a priority to handle such a case in this PR. Let's mark it as a follow-up. Have you come across such cases?

Reply:
We can do it like that.  Then all you have to do is finalize the documentation and parameter names and remove Runit.  I'll take care of it late this evening.  If you want to have it through beforehand, you can of course do it too.

Reply:
Resolving in favor of handling in follow-up: #2343